File: TyreVibes/Core/Component/HeatMapView.swift
//
//  HeatMapView.swift
//  TyreVibes
//
//  Advanced heat map visualization for tyre depth analysis
//

import SwiftUI

// MARK: - Heat Map View
struct HeatMapView: View {
    let heatMap: DepthHeatMap
    let minDepth: Double
    let maxDepth: Double
    @State private var selectedCell: (row: Int, col: Int)?
    @State private var showLegend: Bool = true

    var body: some View {
        VStack(spacing: 16) {
            // Title
            HStack {
                Text("Depth Heat Map")
                    .font(.customFont(size: 18, weight: .bold))
                    .foregroundColor(.white)

                Spacer()

                Button(action: { showLegend.toggle() }) {
                    Image(systemName: showLegend ? "eye.fill" : "eye.slash.fill")
                        .font(.system(size: 16))
                        .foregroundColor(.white.opacity(0.7))
                }
            }
            .padding(.horizontal, 20)

            // Heat Map Grid
            GeometryReader { geometry in
                let cellWidth = geometry.size.width / CGFloat(heatMap.gridSize.columns)
                let cellHeight = geometry.size.height / CGFloat(heatMap.gridSize.rows)

                ZStack {
                    // Grid cells
                    ForEach(0..<heatMap.gridSize.rows, id: \.self) { row in
                        ForEach(0..<heatMap.gridSize.columns, id: \.self) { col in
                            let depth = heatMap.dataPoints[row][col]
                            let color = heatMap.colorForDepth(depth, minDepth: minDepth, maxDepth: maxDepth)

                            Rectangle()
                                .fill(color)
                                .frame(
                                    width: cellWidth - 2,
                                    height: cellHeight - 2
                                )
                                .position(
                                    x: CGFloat(col) * cellWidth + cellWidth / 2,
                                    y: CGFloat(row) * cellHeight + cellHeight / 2
                                )
                                .overlay(
                                    RoundedRectangle(cornerRadius: 4)
                                        .stroke(
                                            selectedCell?.row == row && selectedCell?.col == col
                                                ? Color.white
                                                : Color.clear,
                                            lineWidth: 3
                                        )
                                        .frame(
                                            width: cellWidth - 2,
                                            height: cellHeight - 2
                                        )
                                        .position(
                                            x: CGFloat(col) * cellWidth + cellWidth / 2,
                                            y: CGFloat(row) * cellHeight + cellHeight / 2
                                        )
                                )
                                .onTapGesture {
                                    withAnimation(.spring(response: 0.3, dampingFraction: 0.7)) {
                                        if selectedCell?.row == row && selectedCell?.col == col {
                                            selectedCell = nil
                                        } else {
                                            selectedCell = (row, col)
                                        }
                                    }
                                }
                        }
                    }

                    // Overlay selected cell value
                    if let selected = selectedCell {
                        let depth = heatMap.dataPoints[selected.row][selected.col]
                        let x = CGFloat(selected.col) * cellWidth + cellWidth / 2
                        let y = CGFloat(selected.row) * cellHeight + cellHeight / 2

                        VStack(spacing: 4) {
                            Text(String(format: "%.1f mm", depth))
                                .font(.system(size: 14, weight: .bold))
                                .foregroundColor(.white)
                            Text("Row \(selected.row), Col \(selected.col)")
                                .font(.system(size: 10))
                                .foregroundColor(.white.opacity(0.8))
                        }
                        .padding(8)
                        .background(
                            RoundedRectangle(cornerRadius: 8)
                                .fill(Color.black.opacity(0.8))
                        )
                        .position(x: x, y: max(40, min(y, geometry.size.height - 40)))
                    }
                }
            }
            .aspectRatio(1.5, contentMode: .fit)
            .background(Color.black.opacity(0.2))
            .cornerRadius(12)

            // Legend
            if showLegend {
                HeatMapLegend(
                    colorScheme: heatMap.colorScheme,
                    minValue: minDepth,
                    maxValue: maxDepth
                )
                .transition(.opacity.combined(with: .move(edge: .bottom)))
            }
        }
        .padding(.horizontal, 20)
    }
}

// MARK: - Heat Map Legend
struct HeatMapLegend: View {
    let colorScheme: DepthHeatMap.HeatMapColorScheme
    let minValue: Double
    let maxValue: Double

    var body: some View {
        VStack(spacing: 8) {
            Text("Depth Scale (mm)")
                .font(.customFont(size: 12, weight: .semibold))
                .foregroundColor(.white.opacity(0.8))

            HStack(spacing: 0) {
                // Color gradient bar
                LinearGradient(
                    colors: colorScheme.colors,
                    startPoint: .leading,
                    endPoint: .trailing
                )
                .frame(height: 20)
                .cornerRadius(10)
                .overlay(
                    RoundedRectangle(cornerRadius: 10)
                        .stroke(Color.white.opacity(0.3), lineWidth: 1)
                )
            }

            // Min/Max labels
            HStack {
                Text(String(format: "%.1f", minValue))
                    .font(.customFont(size: 11, weight: .medium))
                    .foregroundColor(.white.opacity(0.7))

                Spacer()

                Text(String(format: "%.1f", (minValue + maxValue) / 2))
                    .font(.customFont(size: 11, weight: .medium))
                    .foregroundColor(.white.opacity(0.7))

                Spacer()

                Text(String(format: "%.1f", maxValue))
                    .font(.customFont(size: 11, weight: .medium))
                    .foregroundColor(.white.opacity(0.7))
            }
        }
        .padding(12)
        .background(
            RoundedRectangle(cornerRadius: 12)
                .fill(Color.black.opacity(0.3))
        )
    }
}

// MARK: - 3D-Style Heat Map (Advanced)
struct HeatMap3DView: View {
    let heatMap: DepthHeatMap
    let minDepth: Double
    let maxDepth: Double
    @State private var rotationAngle: Double = 0
    @State private var elevation: Double = 20

    var body: some View {
        VStack(spacing: 16) {
            Text("3D Depth Visualization")
                .font(.customFont(size: 18, weight: .bold))
                .foregroundColor(.white)
                .padding(.horizontal, 20)

            GeometryReader { geometry in
                Canvas { context, size in
                    let cellWidth = size.width / CGFloat(heatMap.gridSize.columns)
                    let cellHeight = size.height / CGFloat(heatMap.gridSize.rows)

                    for row in 0..<heatMap.gridSize.rows {
                        for col in 0..<heatMap.gridSize.columns {
                            let depth = heatMap.dataPoints[row][col]
                            let color = heatMap.colorForDepth(depth, minDepth: minDepth, maxDepth: maxDepth)

                            // Calculate 3D position
                            let x = CGFloat(col) * cellWidth
                            let y = CGFloat(row) * cellHeight

                            // Elevation based on depth
                            let normalizedDepth = (depth - minDepth) / (maxDepth - minDepth)
                            let z = normalizedDepth * elevation

                            // Apply pseudo-3D transformation
                            let transformedX = x + z * cos(rotationAngle * .pi / 180)
                            let transformedY = y - z * sin(rotationAngle * .pi / 180)

                            let rect = CGRect(
                                x: transformedX,
                                y: transformedY,
                                width: cellWidth - 2,
                                height: cellHeight - 2
                            )

                            context.fill(
                                Path(roundedRect: rect, cornerRadius: 2),
                                with: .color(color)
                            )

                            // Add highlight for depth
                            if normalizedDepth > 0.7 {
                                context.stroke(
                                    Path(roundedRect: rect, cornerRadius: 2),
                                    with: .color(.white.opacity(0.3)),
                                    lineWidth: 1
                                )
                            }
                        }
                    }
                }
            }
            .aspectRatio(1.5, contentMode: .fit)
            .background(Color.black.opacity(0.2))
            .cornerRadius(12)
            .gesture(
                DragGesture()
                    .onChanged { value in
                        rotationAngle = Double(value.translation.width) / 2
                        elevation = 20 + Double(value.translation.height) / 10
                    }
            )

            Text("Drag to rotate and adjust elevation")
                .font(.customFont(size: 12, weight: .medium))
                .foregroundColor(.white.opacity(0.6))
        }
        .padding(.horizontal, 20)
    }
}

// MARK: - Interpolated Heat Map
struct InterpolatedHeatMapView: View {
    let measurements: [DepthMeasurementPoint]
    let gridSize: Int
    let minDepth: Double
    let maxDepth: Double

    private var heatMap: DepthHeatMap {
        generateInterpolatedHeatMap()
    }

    var body: some View {
        HeatMapView(
            heatMap: heatMap,
            minDepth: minDepth,
            maxDepth: maxDepth
        )
    }

    private func generateInterpolatedHeatMap() -> DepthHeatMap {
        var grid: [[Double]] = Array(
            repeating: Array(repeating: 0.0, count: gridSize),
            count: gridSize
        )

        // Interpolate depth values using inverse distance weighting
        for row in 0..<gridSize {
            for col in 0..<gridSize {
                let x = Double(col) / Double(gridSize - 1)
                let y = Double(row) / Double(gridSize - 1)

                let interpolatedDepth = inverseDistanceWeighting(
                    at: CGPoint(x: x, y: y),
                    from: measurements
                )

                grid[row][col] = interpolatedDepth
            }
        }

        return DepthHeatMap(
            gridSize: DepthHeatMap.GridSize(rows: gridSize, columns: gridSize),
            dataPoints: grid,
            colorScheme: .thermal,
            interpolated: true
        )
    }

    private func inverseDistanceWeighting(
        at point: CGPoint,
        from measurements: [DepthMeasurementPoint],
        power: Double = 2.0
    ) -> Double {
        var weightedSum = 0.0
        var weightSum = 0.0

        for measurement in measurements {
            let dx = point.x - measurement.x
            let dy = point.y - measurement.y
            let distance = sqrt(dx * dx + dy * dy)

            // Avoid division by zero
            if distance < 0.001 {
                return measurement.depth
            }

            let weight = 1.0 / pow(distance, power)
            weightedSum += weight * measurement.depth
            weightSum += weight
        }

        return weightSum > 0 ? weightedSum / weightSum : 0
    }
}

// MARK: - Heat Map Statistics Panel
struct HeatMapStatisticsPanel: View {
    let heatMap: DepthHeatMap
    let depthAnalysis: DepthAnalysis

    var body: some View {
        VStack(alignment: .leading, spacing: 16) {
            Text("Statistics")
                .font(.customFont(size: 16, weight: .bold))
                .foregroundColor(.white)

            VStack(spacing: 12) {
                StatRow(label: "Average Depth", value: String(format: "%.2f mm", depthAnalysis.average))
                StatRow(label: "Minimum Depth", value: String(format: "%.2f mm", depthAnalysis.minimum))
                StatRow(label: "Maximum Depth", value: String(format: "%.2f mm", depthAnalysis.maximum))
                StatRow(label: "Std. Deviation", value: String(format: "%.2f mm", depthAnalysis.standardDeviation))
                StatRow(label: "Measurements", value: "\(depthAnalysis.measurements.count)")
            }
            .padding(12)
            .background(
                RoundedRectangle(cornerRadius: 12)
                    .fill(Color.black.opacity(0.3))
            )
        }
        .padding(.horizontal, 20)
    }
}

struct StatRow: View {
    let label: String
    let value: String

    var body: some View {
        HStack {
            Text(label)
                .font(.customFont(size: 13, weight: .medium))
                .foregroundColor(.white.opacity(0.7))

            Spacer()

            Text(value)
                .font(.customFont(size: 13, weight: .semibold))
                .foregroundColor(.white)
        }
    }
}

File: android/app/src/main/java/it/tyrevibes/app/ui/components/HeatMapView.kt
package it.tyrevibes.app.ui.components

import androidx.compose.animation.core.*
import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Visibility
import androidx.compose.material.icons.filled.VisibilityOff
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.CornerRadius
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import it.tyrevibes.app.ui.theme.SoraFontFamily
import kotlin.math.pow
import kotlin.math.sqrt

/**
 * Heat Map Data Models
 */
data class DepthHeatMap(
    val gridSize: GridSize,
    val dataPoints: List<List<Double>>,
    val colorScheme: HeatMapColorScheme,
    val interpolated: Boolean = false
) {
    data class GridSize(val rows: Int, val columns: Int)

    enum class HeatMapColorScheme {
        THERMAL, RAINBOW, MONOCHROME;

        val colors: List<Color>
            get() = when (this) {
                THERMAL -> listOf(
                    Color(0xFF00FF00), // Green (good)
                    Color(0xFFFFFF00), // Yellow
                    Color(0xFFFF8C00), // Orange
                    Color(0xFFFF0000)  // Red (critical)
                )
                RAINBOW -> listOf(
                    Color(0xFF0000FF), // Blue
                    Color(0xFF00FFFF), // Cyan
                    Color(0xFF00FF00), // Green
                    Color(0xFFFFFF00), // Yellow
                    Color(0xFFFF0000)  // Red
                )
                MONOCHROME -> listOf(
                    Color(0xFF000000),
                    Color(0xFF808080),
                    Color(0xFFFFFFFF)
                )
            }
    }

    fun colorForDepth(depth: Double, minDepth: Double, maxDepth: Double): Color {
        val normalized = ((depth - minDepth) / (maxDepth - minDepth)).coerceIn(0.0, 1.0)
        val colors = colorScheme.colors
        val index = (normalized * (colors.size - 1)).toFloat()
        val lowerIndex = index.toInt().coerceIn(0, colors.size - 2)
        val upperIndex = (lowerIndex + 1).coerceIn(0, colors.size - 1)
        val fraction = index - lowerIndex

        return lerp(colors[lowerIndex], colors[upperIndex], fraction)
    }

    private fun lerp(start: Color, end: Color, fraction: Float): Color {
        return Color(
            red = start.red + (end.red - start.red) * fraction,
            green = start.green + (end.green - start.green) * fraction,
            blue = start.blue + (end.blue - start.blue) * fraction,
            alpha = start.alpha + (end.alpha - start.alpha) * fraction
        )
    }
}

data class DepthMeasurementPoint(
    val x: Double,
    val y: Double,
    val depth: Double,
    val timestamp: Long = System.currentTimeMillis()
)

/**
 * Heat Map View Component
 */
@Composable
fun HeatMapView(
    heatMap: DepthHeatMap,
    minDepth: Double,
    maxDepth: Double,
    modifier: Modifier = Modifier
) {
    var selectedCell by remember { mutableStateOf<Pair<Int, Int>?>(null) }
    var showLegend by remember { mutableStateOf(true) }

    Column(
        modifier = modifier
            .fillMaxWidth()
            .padding(horizontal = 20.dp),
        verticalArrangement = Arrangement.spacedBy(16.dp)
    ) {
        // Header
        Row(
            modifier = Modifier.fillMaxWidth(),
            horizontalArrangement = Arrangement.SpaceBetween,
            verticalAlignment = Alignment.CenterVertically
        ) {
            Text(
                text = "Mappa di Profondità",
                fontFamily = SoraFontFamily,
                fontWeight = FontWeight.Bold,
                fontSize = 18.sp,
                color = Color.White
            )

            IconButton(onClick = { showLegend = !showLegend }) {
                Icon(
                    imageVector = if (showLegend) Icons.Default.Visibility else Icons.Default.VisibilityOff,
                    contentDescription = "Toggle legend",
                    tint = Color.White.copy(alpha = 0.7f)
                )
            }
        }

        // Heat Map Grid
        Box(
            modifier = Modifier
                .fillMaxWidth()
                .aspectRatio(1.5f)
                .background(
                    color = Color.Black.copy(alpha = 0.2f),
                    shape = RoundedCornerShape(12.dp)
                )
        ) {
            Canvas(
                modifier = Modifier
                    .fillMaxSize()
                    .padding(8.dp)
            ) {
                val cellWidth = size.width / heatMap.gridSize.columns
                val cellHeight = size.height / heatMap.gridSize.rows

                for (row in 0 until heatMap.gridSize.rows) {
                    for (col in 0 until heatMap.gridSize.columns) {
                        val depth = heatMap.dataPoints[row][col]
                        val color = heatMap.colorForDepth(depth, minDepth, maxDepth)

                        val isSelected = selectedCell?.first == row && selectedCell?.second == col

                        drawRoundRect(
                            color = color,
                            topLeft = Offset(col * cellWidth + 1.dp.toPx(), row * cellHeight + 1.dp.toPx()),
                            size = Size(cellWidth - 2.dp.toPx(), cellHeight - 2.dp.toPx()),
                            cornerRadius = CornerRadius(4.dp.toPx())
                        )

                        // Selected cell border
                        if (isSelected) {
                            drawRoundRect(
                                color = Color.White,
                                topLeft = Offset(col * cellWidth, row * cellHeight),
                                size = Size(cellWidth, cellHeight),
                                cornerRadius = CornerRadius(4.dp.toPx()),
                                style = androidx.compose.ui.graphics.drawscope.Stroke(width = 3.dp.toPx())
                            )
                        }
                    }
                }
            }

            // Selected cell info overlay
            selectedCell?.let { (row, col) ->
                val depth = heatMap.dataPoints[row][col]
                Box(
                    modifier = Modifier
                        .align(Alignment.TopCenter)
                        .padding(top = 16.dp)
                        .background(
                            color = Color.Black.copy(alpha = 0.8f),
                            shape = RoundedCornerShape(8.dp)
                        )
                        .padding(8.dp)
                ) {
                    Column(
                        horizontalAlignment = Alignment.CenterHorizontally,
                        verticalArrangement = Arrangement.spacedBy(4.dp)
                    ) {
                        Text(
                            text = String.format("%.1f mm", depth),
                            fontFamily = SoraFontFamily,
                            fontWeight = FontWeight.Bold,
                            fontSize = 14.sp,
                            color = Color.White
                        )
                        Text(
                            text = "Row $row, Col $col",
                            fontFamily = SoraFontFamily,
                            fontSize = 10.sp,
                            color = Color.White.copy(alpha = 0.8f)
                        )
                    }
                }
            }
        }

        // Legend
        if (showLegend) {
            HeatMapLegend(
                colorScheme = heatMap.colorScheme,
                minValue = minDepth,
                maxValue = maxDepth
            )
        }
    }
}

@Composable
private fun HeatMapLegend(
    colorScheme: DepthHeatMap.HeatMapColorScheme,
    minValue: Double,
    maxValue: Double,
    modifier: Modifier = Modifier
) {
    Column(
        modifier = modifier
            .fillMaxWidth()
            .background(
                color = Color.Black.copy(alpha = 0.3f),
                shape = RoundedCornerShape(12.dp)
            )
            .padding(12.dp),
        verticalArrangement = Arrangement.spacedBy(8.dp)
    ) {
        Text(
            text = "Scala Profondità (mm)",
            fontFamily = SoraFontFamily,
            fontWeight = FontWeight.SemiBold,
            fontSize = 12.sp,
            color = Color.White.copy(alpha = 0.8f)
        )

        // Gradient bar
        Box(
            modifier = Modifier
                .fillMaxWidth()
                .height(20.dp)
                .background(
                    brush = Brush.horizontalGradient(
                        colors = colorScheme.colors
                    ),
                    shape = RoundedCornerShape(10.dp)
                )
        )

        // Min/Max labels
        Row(
            modifier = Modifier.fillMaxWidth(),
            horizontalArrangement = Arrangement.SpaceBetween
        ) {
            Text(
                text = String.format("%.1f", minValue),
                fontFamily = SoraFontFamily,
                fontWeight = FontWeight.Medium,
                fontSize = 11.sp,
                color = Color.White.copy(alpha = 0.7f)
            )

            Text(
                text = String.format("%.1f", (minValue + maxValue) / 2),
                fontFamily = SoraFontFamily,
                fontWeight = FontWeight.Medium,
                fontSize = 11.sp,
                color = Color.White.copy(alpha = 0.7f)
            )

            Text(
                text = String.format("%.1f", maxValue),
                fontFamily = SoraFontFamily,
                fontWeight = FontWeight.Medium,
                fontSize = 11.sp,
                color = Color.White.copy(alpha = 0.7f)
            )
        }
    }
}

/**
 * Interpolated Heat Map View - genera heat map da punti sparsi
 */
@Composable
fun InterpolatedHeatMapView(
    measurements: List<DepthMeasurementPoint>,
    gridSize: Int = 10,
    minDepth: Double,
    maxDepth: Double,
    modifier: Modifier = Modifier
) {
    val heatMap = remember(measurements, gridSize) {
        generateInterpolatedHeatMap(measurements, gridSize)
    }

    HeatMapView(
        heatMap = heatMap,
        minDepth = minDepth,
        maxDepth = maxDepth,
        modifier = modifier
    )
}

private fun generateInterpolatedHeatMap(
    measurements: List<DepthMeasurementPoint>,
    gridSize: Int
): DepthHeatMap {
    val grid = MutableList(gridSize) { MutableList(gridSize) { 0.0 } }

    for (row in 0 until gridSize) {
        for (col in 0 until gridSize) {
            val x = col.toDouble() / (gridSize - 1)
            val y = row.toDouble() / (gridSize - 1)

            grid[row][col] = inverseDistanceWeighting(
                x = x,
                y = y,
                measurements = measurements,
                power = 2.0
            )
        }
    }

    return DepthHeatMap(
        gridSize = DepthHeatMap.GridSize(rows = gridSize, columns = gridSize),
        dataPoints = grid,
        colorScheme = DepthHeatMap.HeatMapColorScheme.THERMAL,
        interpolated = true
    )
}

private fun inverseDistanceWeighting(
    x: Double,
    y: Double,
    measurements: List<DepthMeasurementPoint>,
    power: Double = 2.0
): Double {
    var weightedSum = 0.0
    var weightSum = 0.0

    for (measurement in measurements) {
        val dx = x - measurement.x
        val dy = y - measurement.y
        val distance = sqrt(dx * dx + dy * dy)

        // Avoid division by zero
        if (distance < 0.001) {
            return measurement.depth
        }

        val weight = 1.0 / distance.pow(power)
        weightedSum += weight * measurement.depth
        weightSum += weight
    }

    return if (weightSum > 0) weightedSum / weightSum else 0.0
}

/**
 * Heat Map Statistics Panel
 */
@Composable
fun HeatMapStatisticsPanel(
    heatMap: DepthHeatMap,
    averageDepth: Double,
    minDepth: Double,
    maxDepth: Double,
    standardDeviation: Double,
    measurementCount: Int,
    modifier: Modifier = Modifier
) {
    Column(
        modifier = modifier
            .fillMaxWidth()
            .padding(horizontal = 20.dp)
            .background(
                color = Color.Black.copy(alpha = 0.3f),
                shape = RoundedCornerShape(12.dp)
            )
            .padding(12.dp),
        verticalArrangement = Arrangement.spacedBy(12.dp)
    ) {
        Text(
            text = "Statistiche",
            fontFamily = SoraFontFamily,
            fontWeight = FontWeight.Bold,
            fontSize = 16.sp,
            color = Color.White
        )

        Column(
            verticalArrangement = Arrangement.spacedBy(12.dp)
        ) {
            StatRow("Profondità Media", String.format("%.2f mm", averageDepth))
            StatRow("Profondità Minima", String.format("%.2f mm", minDepth))
            StatRow("Profondità Massima", String.format("%.2f mm", maxDepth))
            StatRow("Deviazione Standard", String.format("%.2f mm", standardDeviation))
            StatRow("Misurazioni", measurementCount.toString())
        }
    }
}

@Composable
private fun StatRow(label: String, value: String) {
    Row(
        modifier = Modifier.fillMaxWidth(),
        horizontalArrangement = Arrangement.SpaceBetween
    ) {
        Text(
            text = label,
            fontFamily = SoraFontFamily,
            fontWeight = FontWeight.Medium,
            fontSize = 13.sp,
            color = Color.White.copy(alpha = 0.7f)
        )

        Text(
            text = value,
            fontFamily = SoraFontFamily,
            fontWeight = FontWeight.SemiBold,
            fontSize = 13.sp,
            color = Color.White
        )
    }
}
